sopel/tests/pytest_plugin.py

```diff
@@ -5,11 +5,163 @@
+
+        match = parse_results[0]
+        sender = mockbot.nick if privmsg else "#channel"
+        hostmask = "%s!%s@%s" % (mockbot.nick, "UserName", "example.com")
+
+        # TODO enable message tags
+        full_message = ':{} PRIVMSG {} :{}'.format(hostmask, sender, msg)
+        pretrigger = trigger.PreTrigger(
+            mockbot.nick, full_message, url_schemes=url_schemes)
+        test_trigger = trigger.Trigger(mockbot.settings, pretrigger, match)
+        pattern = re.compile(r'^%s: ' % re.escape(mockbot.nick))
+
+        # setup module
+        module = sys.modules[tested_func.__module__]
+        if hasattr(module, 'setup'):
+            module.setup(mockbot)
+
+        def isnt_ignored(value):
+            """Return True if value doesn't match any re in ignore list."""
+            return not any(
+                re.match(ignored_line, value)
+                for ignored_line in ignore)
+
+        expected_output_count = 0
+        for _i in range(repeat):
+            expected_output_count += len(results)
+            wrapper = bot.SopelWrapper(mockbot, test_trigger)
+            tested_func(wrapper, test_trigger)
+
+            output_triggers = (
+                trigger.PreTrigger(
+                    mockbot.nick,
+                    message.decode('utf-8'),
+                    url_schemes=url_schemes,
+                )
+                for message in wrapper.backend.message_sent
+            )
+            output_texts = (
+                # subtract "Sopel: " when necessary
+                pattern.sub('', output_trigger.args[-1])
+                for output_trigger in output_triggers
+            )
+            outputs = [text for text in output_texts if isnt_ignored(text)]
+
+            # output length
+            assert len(outputs) == expected_output_count
+
+            # output content
+            for expected, output in zip(results, outputs):
+                if use_regexp:
+                    message = (
+                        "Output does not match the regex:\n"
+                        "Pattern: %s\n"
+                        "Output: %s"
+                    ) % (expected, output)
+                    if not re.match(expected, output):
+                        raise AssertionError(message)
+                else:
+                    assert expected == output
+
+    return test
+
+
+def insert_into_module(func, module_name, base_name, prefix):
+    """Add a function into a module.
+
+    This can be used to add a test function, a setup function, or a fixture
+    to an existing module to be used with pytest.
+    """
+    func.__module__ = module_name
+    module = sys.modules[module_name]
+    # Make sure the func method does not overwrite anything.
+    for i in range(1000):
+        func.__name__ = str("%s_%s_%s" % (prefix, base_name, i))
+        if not hasattr(module, func.__name__):
+            break
+    setattr(module, func.__name__, func)
```

Comment:
I know this bug is from the original code that you just cut-and-pasted here, but it really annoys me that there's no exception raised if some insane plugin already has 1000 conflicting objects and the code will simply overwrite the last one (`prefix_basename_999`) if it hasn't found a free slot yet.

Obviously this isn't a practical issue ~~except for plugins _you_ write 😝~~ but still.